File: system/public/aesthetic.computer/lib/arena-world.mjs
// arena-world.mjs — shared static arena geometry.
// Imported by both disks/arena.mjs (client render + prediction) and
// session-server/arena-manager.mjs (authoritative collision). Keep this
// dep-free so it works in Node and the browser.
//
// Coordinates are world-space. Y is up, ground sits at -1.5. Obstacles are
// flat-top extrusions: pillars are cylinders, walls are axis-aligned boxes.
// Heights are tuned so a stand-jump can clear the platform edge but not the
// obstacle tops — strafe-jumping skill earns elevation.

const GROUND_Y = -1.5;

// Horizontal extents of obstacle layout. Players can run between them.
// Pillars are placed in the four "shoulders" of the arena, walls form a
// central S-curve so cover lines aren't symmetric.
export const ARENA_OBSTACLES = Object.freeze([
  // Four corner pillars (cylinders). r=0.8, h=4.
  { type: "cylinder", x:  8, z:  8, r: 0.8, yMin: GROUND_Y, yMax: GROUND_Y + 4 },
  { type: "cylinder", x: -8, z:  8, r: 0.8, yMin: GROUND_Y, yMax: GROUND_Y + 4 },
  { type: "cylinder", x:  8, z: -8, r: 0.8, yMin: GROUND_Y, yMax: GROUND_Y + 4 },
  { type: "cylinder", x: -8, z: -8, r: 0.8, yMin: GROUND_Y, yMax: GROUND_Y + 4 },

  // Center totem: tall thin pillar at origin.
  { type: "cylinder", x: 0, z: 0, r: 0.5, yMin: GROUND_Y, yMax: GROUND_Y + 5 },

  // Two staggered cover walls forming a soft S-curve. Thickness 0.6, height
  // 2.4 — tall enough to break sightlines, short enough that strafe-jumps
  // can peek over the top.
  {
    type: "box",
    xMin: -5.5, xMax:  -0.5,
    zMin: -3.3, zMax:  -2.7,
    yMin: GROUND_Y, yMax: GROUND_Y + 2.4,
  },
  {
    type: "box",
    xMin:  0.5, xMax:  5.5,
    zMin:  2.7, zMax:  3.3,
    yMin: GROUND_Y, yMax: GROUND_Y + 2.4,
  },
]);

// Per-obstacle visual color palette (paint helpers; not used by physics).
// Indices match ARENA_OBSTACLES order so render code can pick per-shape tints.
export const ARENA_OBSTACLE_COLORS = Object.freeze([
  [0.55, 0.42, 0.30, 1.0],   // NE pillar — warm sand
  [0.42, 0.50, 0.55, 1.0],   // NW pillar — cool slate
  [0.48, 0.40, 0.45, 1.0],   // SE pillar — dusty plum
  [0.40, 0.46, 0.36, 1.0],   // SW pillar — moss
  [0.62, 0.55, 0.32, 1.0],   // center totem — brassy ochre
  [0.38, 0.34, 0.42, 1.0],   // wall A — deep indigo-stone
  [0.44, 0.36, 0.32, 1.0],   // wall B — terracotta
]);

// 🏃 Quake-style movement tuning for the arena (shared client/server).
export const ARENA_PHYSICS = Object.freeze({
  airAccel: 70,        // u/s² along wishdir while airborne
  groundAccel: 80,     // u/s² along wishdir on the ground
  airCapSpeed: 1.5,    // Q3 air-control window
  groundFriction: 6,
  playerRadius: 0.4,
});

File: arena-native/Sources/ArenaNative/ArenaWorld.swift
// ArenaWorld.swift — Swift port of lib/arena-world.mjs. Static arena geometry
// shared by physics (collision) and rendering. World-space; Y is up; the
// ground sits at -1.5. Obstacles are flat-top extrusions: pillars are
// cylinders, walls are axis-aligned boxes.

import Foundation

enum ObstacleKind { case box, cylinder }

// Unified obstacle. Box uses xMin..zMax; cylinder uses x,z,r. Both carry an
// optional vertical span [yMin, yMax].
struct Obstacle {
    let kind: ObstacleKind
    // box bounds
    var xMin: Double = 0, xMax: Double = 0, zMin: Double = 0, zMax: Double = 0
    // cylinder center + radius
    var x: Double = 0, z: Double = 0, r: Double = 0
    var yMin: Double?
    var yMax: Double?

    static func cylinder(x: Double, z: Double, r: Double, yMin: Double, yMax: Double) -> Obstacle {
        Obstacle(kind: .cylinder, x: x, z: z, r: r, yMin: yMin, yMax: yMax)
    }
    static func box(xMin: Double, xMax: Double, zMin: Double, zMax: Double,
                    yMin: Double, yMax: Double) -> Obstacle {
        Obstacle(kind: .box, xMin: xMin, xMax: xMax, zMin: zMin, zMax: zMax,
                 yMin: yMin, yMax: yMax)
    }
}

enum ArenaWorld {
    static let groundY: Double = -1.5

    // ARENA_OBSTACLES — four corner pillars, a center totem, two S-curve walls.
    static let obstacles: [Obstacle] = [
        .cylinder(x:  8, z:  8, r: 0.8, yMin: groundY, yMax: groundY + 4),
        .cylinder(x: -8, z:  8, r: 0.8, yMin: groundY, yMax: groundY + 4),
        .cylinder(x:  8, z: -8, r: 0.8, yMin: groundY, yMax: groundY + 4),
        .cylinder(x: -8, z: -8, r: 0.8, yMin: groundY, yMax: groundY + 4),
        .cylinder(x:  0, z:  0, r: 0.5, yMin: groundY, yMax: groundY + 5),
        .box(xMin: -5.5, xMax: -0.5, zMin: -3.3, zMax: -2.7, yMin: groundY, yMax: groundY + 2.4),
        .box(xMin:  0.5, xMax:  5.5, zMin:  2.7, zMax:  3.3, yMin: groundY, yMax: groundY + 2.4),
    ]

    // ARENA_OBSTACLE_COLORS — index-matched per-obstacle tints (RGB 0..1).
    static let obstacleColors: [SIMD3<Float>] = [
        SIMD3(0.55, 0.42, 0.30), // NE pillar — warm sand
        SIMD3(0.42, 0.50, 0.55), // NW pillar — cool slate
        SIMD3(0.48, 0.40, 0.45), // SE pillar — dusty plum
        SIMD3(0.40, 0.46, 0.36), // SW pillar — moss
        SIMD3(0.62, 0.55, 0.32), // center totem — brassy ochre
        SIMD3(0.38, 0.34, 0.42), // wall A — deep indigo-stone
        SIMD3(0.44, 0.36, 0.32), // wall B — terracotta
    ]

    // groundBounds (matches ARENA_CFG).
    static let bounds = (xMin: -14.0, xMax: 14.0, zMin: -14.0, zMax: 14.0)
}
